File: ruby/perceptron.rb
class Perceptron

  def to_train(train_data, target, error_threshold=0.1)
    @error_threshold = error_threshold
    @train_data = train_data
    @target = target
    @weights = Array.new(train_data.first.size + 1, 1)

    loop do
      it_is_over = true
      @target.size.times do |line|
        error = calculate_error(line)
        unless error == 0
          calculate_weights(error, line)
          it_is_over = false
        end
      end

      break if it_is_over
    end
  end

  def calculate_output(data)
    total = weights[0]
    data.each_with_index do |value, index|
      total += value * weights[index+1]
    end
    (total >= 0) ? 1 : 0
  end

  private
    attr_accessor :weights, :error_threshold, :train_data, :target

    def calculate_error(line)
      target[line] - calculate_output(train_data[line])
    end

    def calculate_weights(error, line)
      weights[0] = weights[0] + error_threshold * error
      train_data[line].each_with_index do |value, index|
        weights[index+1] = weights[index+1] + error_threshold * error * value
      end
    end
end

File: python/perceptron.py
class Perceptron(object):

	def to_train(self, trainData, target, errorThreshold=0.1):
		self.errorThreshold = errorThreshold
		self.trainData = trainData
		self.target = target
		self.weights = [1]
		for i in range(len(trainData[0])):
			self.weights = self.weights + [1]

		it_is_over = False
		while not it_is_over:
			it_is_over = True
			for line in range(len(self.target)):
				error = self.calculateError(line)
				if (error != 0):
					self.calculateWeights(error, line)
					it_is_over = False

	def calculateOutput(self, data):
		total = self.weights[0]
		for index in range(len(data)):
			total += data[index] * self.weights[index+1]

		return 1 if(total >= 0) else 0

	def calculateError(self, line):
		return self.target[line] - self.calculateOutput(self.trainData[line])

	def calculateWeights(self, error, line):
		self.weights[0] = self.weights[0] + self.errorThreshold * error
		for index in range(len(self.trainData[line])):
			self.weights[index+1] = self.weights[index+1] + self.errorThreshold * error * self.trainData[line][index]